pkg/bpf/events.go

```diff
@@ -0,0 +1,276 @@
+	v := h.closed.Load()
+	return v.(bool)
+}
+
+func (h *Handle) isFull() bool {
+	return len(h.c) >= cap(h.c)
+}
+
+// This configures how big buffers are for channels used for streaming events from
+// eventsBuffer.
+//
+// To prevent blocking bpf.Map operations, subscribed events are buffered per client handle.
+// How fast subscribers will need to proceess events will depend on the event throughput.
+// In this case, our throughput will be expected to be not above 100 events a second.
+// Therefore the consumer will have 10ms to process each event. The channel is also
+// given a constant buffer size in the case where events arrive at once (i.e. all 100 events
+// arriving at the top of the second).
+//
+// NOTE: Although using timers/timed-contexts seems like an obvious choice for this use case,
+// the timer.After implementation actually uses a large amount of memory. To reduce memory spikes
+// in high throughput cases, we instead just use a sufficiently buffered channel.
+const (
+	eventSubChanBufferSize = 256
+)
+
+func (eb *eventsBuffer) dumpAndSubscribe(callback EventCallbackFunc, follow bool) *Handle {
+	if callback != nil {
+		eb.dumpWithCallback(callback)
+	}
+
+	if !follow {
+		return nil
+	}
+
+	closed := &atomic.Value{}
+	closed.Store(false)
+	h := &Handle{
+		c:      make(chan *Event, eventSubChanBufferSize),
+		closer: &sync.Once{},
+		closed: closed,
+	}
+
+	eb.subsLock.Lock()
+	defer eb.subsLock.Unlock()
+	eb.subscriptions = append(eb.subscriptions, h)
+	return h
+}
+
+// DumpAndSubscribe dumps existing buffer, if callback is not nil. Followed by creating a
+// subscription to the maps events buffer and returning the handle.
+// These actions are done together so as to prevent possible missed events between the handoff
+// of the callback and sub handle creation.
+func (m *Map) DumpAndSubscribe(callback EventCallbackFunc, follow bool) *Handle {
+	// note: we have to hold rlock for the duration of this to prevent missed events between dump and sub.
+	// dumpAndSubscribe maintains its own write-lock for updating subscribers.
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	if !m.eventsBufferEnabled {
+		return nil
+	}
+	return m.events.dumpAndSubscribe(callback, follow)
+}
+
+func (m *Map) IsEventsEnabled() bool {
+	return m.eventsBufferEnabled
+}
+
+func (eb *eventsBuffer) add(e *Event) {
+	eb.buffer.Add(e)
+	var activeSubs []*Handle
+	activeSubsLock := &lock.Mutex{}
+	wg := &sync.WaitGroup{}
+	for i, sub := range eb.subscriptions {
+		if sub.isClosed() { // sub will be removed.
+			continue
+		}
+		wg.Add(1)
+		go func(sub *Handle, i int) {
+			defer wg.Done()
+			if sub.isFull() {
+				err := fmt.Errorf("timed out waiting to send sub map event")
+				log.WithError(err).Warnf("subscription channel buffer %d was full, closing subscription", i)
+				sub.close(err)
+			} else {
+				sub.c <- e
+				activeSubsLock.Lock()
+				activeSubs = append(activeSubs, sub)
+				activeSubsLock.Unlock()
+			}
+		}(sub, i)
+	}
```

Comment:
How many goroutines could be potentially be spinning up here at once?

Reply:
Unbounded, but practically it will be roughly "the number of clients doing --follow on events"

That's good question though, I could introduce some kind of limit to this, or serialize it into batches. However what are the circumstances under which this would cause problems and does it make sense to implement those limits here. 

I could see this causing issues if someone execed in and spam a bunch of forked `cilium map events ... --follow` but I'm sure many such calls would cause problems under this circumstance. How defensive do we want to be with something like that?

If this is something we're concerned about - I think it makes more sense to add that to the http server impl level by adding some kind of rate limit/max concurrent calls to the endpoint.

lmkwyt

Reply:
> Unbounded, but practically it will be bounded by "the number of concurrent clients doing --follow on events"

:ok_hand: 

I think that's good enough for me IMO. I think leaving a comment saying exactly that before the goroutine code tells future readers that they don't need to necessarily worry about this or if someone comes along and notices this assumption somehow no longer holds due to the expectation that's been explicitly documented. Assumptions / expectations are extremely valuable to document in the code base for our future selves; sometimes it can be difficult to assess whether a snippet of code is problematic, this comment would alleviate that..

Ideally for resource constraints particularly CPU / goroutines, we use something like https://github.com/cilium/workerpool in general to avoid unboundedness. We've seen issues before with memory usage and scheduler saturation in hot spots of the code which don't limit the number of goroutines.

Reply:
sounds good,

no matter the implementation, if someone starts opening too many follow commands this would cause issues. I think the correct thing would be to just limit the number of concurrent subs allowed at the same time.

Reply:
actually, I ended up adding a limit on the number of subs. probably much simpler in the long run and it'll be easier to solve any potential issues.